source/databricks/package/balance_fixing_total_production.py

```diff
@@ -0,0 +1,295 @@
+            "body.MeteringPointId",
+            "body.MeteringPointType",
+            "body.GsrnNumber",
+            "body.GridAreaLinkId",
+            "body.ConnectionState",
+            "body.EffectiveDate",
+            "body.Resolution",
+        )
+    )
+
+    window = Window.partitionBy("MeteringPointId").orderBy("EffectiveDate")
+
+    metering_point_periods_df = (
+        metering_point_events_df.withColumn(
+            "toEffectiveDate",
+            lead("EffectiveDate", 1, "2099-01-01T23:00:00.000+0000").over(window),
+        )
+        .withColumn(
+            "GridAreaLinkId",
+            coalesce(col("GridAreaLinkId"), last("GridAreaLinkId", True).over(window)),
+        )
+        .withColumn(
+            "ConnectionState",
+            when(
+                col("MessageType") == "MeteringPointCreated",
+                lit(ConnectionState.new),
+            ).when(
+                col("MessageType") == "MeteringPointConnected",
+                lit(ConnectionState.connected),
+            ),
+        )
+        .withColumn(
+            "MeteringPointType",
+            coalesce(
+                col("MeteringPointType"), last("MeteringPointType", True).over(window)
+            ),
+        )
+        .withColumn(
+            "Resolution",
+            coalesce(col("Resolution"), last("Resolution", True).over(window)),
+        )
+        .where(col("EffectiveDate") <= period_end_datetime)
+        .where(col("toEffectiveDate") >= period_start_datetime)
+        .where(
+            col("ConnectionState") == ConnectionState.connected
+        )  # Only aggregate when metering points is connected
+        .where(col("MeteringPointType") == MeteringPointType.production)
+    )
+
+    # Only include metering points in the selected grid areas
+    metering_point_periods_df = metering_point_periods_df.join(
+        grid_area_df,
+        metering_point_periods_df["GridAreaLinkId"] == grid_area_df["GridAreaLinkId"],
+        "inner",
+    ).select(
+        metering_point_periods_df["MessageType"],
+        "GsrnNumber",
+        "GridAreaCode",
+        "EffectiveDate",
+        "toEffectiveDate",
+        "Resolution",
+    )
+
+    return metering_point_periods_df
+
+
+def _get_enriched_time_series_points_df(
+    raw_time_series_points,
+    metering_point_period_df,
+    snapshot_datetime,
+    period_start_datetime,
+    period_end_datetime,
+) -> DataFrame:
+    timeseries_df = (
+        raw_time_series_points.where(col("storedTime") <= snapshot_datetime)
+        .where(col("time") >= period_start_datetime)
+        .where(col("time") < period_end_datetime)
+        # Quantity of time series points should have 3 digits. Calculations, however, must use 6 digit precision to reduce rounding errors
+        .withColumn("quantity", col("quantity").cast("decimal(18,6)"))
+    )
+
+    # Only use latest registered points
+    window = Window.partitionBy("metering_point_id", "time").orderBy(
```

Comment:
```suggestion
    window = Window.partitionBy("GsrnNumber", "time").orderBy(
```
Should be changed to "GsrnNumber" when [this PR](https://github.com/Energinet-DataHub/geh-timeseries/pull/511) is in main

Reply:
We can do it now as this PR will not change the job in production.